# Week 7: Groupby
wil be looking at the drinks dataset again :)

In [1]:
import pandas as pd

In [2]:
movies_df = pd.read_csv('https://raw.githubusercontent.com/stiles/usc/main/data/raw/imdb_1000.csv')

#### basic information about the dataset

In [3]:
movies_df.head()

star_rating                     title content_rating   genre  duration  \
0          9.3  The Shawshank Redemption              R   Crime       142   
1          9.2             The Godfather              R   Crime       175   
2          9.1    The Godfather: Part II              R   Crime       200   
3          9.0           The Dark Knight          PG-13  Action       152   
4          8.9              Pulp Fiction              R   Crime       154   

                                         actors_list  
0  [u'Tim Robbins', u'Morgan Freeman', u'Bob Gunt...  
1    [u'Marlon Brando', u'Al Pacino', u'James Caan']  
2  [u'Al Pacino', u'Robert De Niro', u'Robert Duv...  
3  [u'Christian Bale', u'Heath Ledger', u'Aaron E...  
4  [u'John Travolta', u'Uma Thurman', u'Samuel L....

In [4]:
movies_df.tail()

star_rating                                            title  \
974          7.4                                          Tootsie   
975          7.4                      Back to the Future Part III   
976          7.4  Master and Commander: The Far Side of the World   
977          7.4                                      Poltergeist   
978          7.4                                      Wall Street   

    content_rating      genre  duration  \
974             PG     Comedy       116   
975             PG  Adventure       118   
976          PG-13     Action       138   
977             PG     Horror       114   
978              R      Crime       126   

                                           actors_list  
974  [u'Dustin Hoffman', u'Jessica Lange', u'Teri G...  
975  [u'Michael J. Fox', u'Christopher Lloyd', u'Ma...  
976  [u'Russell Crowe', u'Paul Bettany', u'Billy Bo...  
977  [u'JoBeth Williams', u"Heather O'Rourke", u'Cr...  
978  [u'Charlie Sheen', u'Michael Douglas', u'Tamar...

##### = data is listed in numerical order of ratings

In [5]:
len(movies_df)

979

##### = 979 movies in the dataset

#### let's explore the dataset

##### 1. which genre has the most movies in the top 1,000?

In [6]:
movies_df.groupby(['genre'])['content_rating'].count()

genre
Action       135
Adventure     74
Animation     62
Biography     76
Comedy       156
Crime        124
Drama        278
Family         2
Fantasy        1
Film-Noir      3
History        1
Horror        29
Mystery       16
Sci-Fi         5
Thriller       5
Western        9
Name: content_rating, dtype: int64

##### = drama category has the highest representation

##### 2. how many are action movies?

In [8]:
len(movies_df[movies_df['genre'] == 'Action'])

136

##### 3. movies from highest to lowest rating

In [13]:
movies_df.sort_values('star_rating', ascending=False)

star_rating                     title content_rating   genre  duration  \
0            9.3  The Shawshank Redemption              R   Crime       142   
1            9.2             The Godfather              R   Crime       175   
2            9.1    The Godfather: Part II              R   Crime       200   
3            9.0           The Dark Knight          PG-13  Action       152   
4            8.9              Pulp Fiction              R   Crime       154   
..           ...                       ...            ...     ...       ...   
948          7.4                Frances Ha              R  Comedy        86   
949          7.4                Home Alone             PG  Comedy       103   
950          7.4                     Bound              R   Crime       108   
951          7.4             Sleepy Hollow              R   Drama       105   
978          7.4               Wall Street              R   Crime       126   

                                           actors_list  
0    [u'Tim Robbins', u'Morgan Freeman', u'Bob Gunt...  
1      [u'Marlon Brando', u'Al Pacino', u'James Caan']  
2    [u'Al Pacino', u'Robert De Niro', u'Robert Duv...  
3    [u'Christian Bale', u'Heath Ledger', u'Aaron E...  
4    [u'John Travolta', u'Uma Thurman', u'Samuel L....  
..                                                 ...  
948  [u'Greta Gerwig', u'Mickey Sumner', u'Adam Dri...  
949  [u'Macaulay Culkin', u'Joe Pesci', u'Daniel St...  
950  [u'Jennifer Tilly', u'Gina Gershon', u'Joe Pan...  
951  [u'Johnny Depp', u'Christina Ricci', u'Miranda...  
978  [u'Charlie Sheen', u'Michael Douglas', u'Tamar...  

[979 rows x 6 columns]

##### 4. movies ranked lower than 7

In [18]:
movies_df[movies_df['star_rating'] < 7]

Empty DataFrame
Columns: [star_rating, title, content_rating, genre, duration, actors_list]
Index: []

##### = none ranked below 7

##### 5. movies ranked above 9

In [20]:
movies_df[movies_df['star_rating'] > 9]

star_rating                     title content_rating  genre  duration  \
0          9.3  The Shawshank Redemption              R  Crime       142   
1          9.2             The Godfather              R  Crime       175   
2          9.1    The Godfather: Part II              R  Crime       200   

                                         actors_list  
0  [u'Tim Robbins', u'Morgan Freeman', u'Bob Gunt...  
1    [u'Marlon Brando', u'Al Pacino', u'James Caan']  
2  [u'Al Pacino', u'Robert De Niro', u'Robert Duv...

##### 6. average movie running time based on genre

In [21]:
movies_df.groupby(['genre']).mean().reset_index()

genre  star_rating    duration
0      Action     7.884559  126.485294
1   Adventure     7.933333  134.840000
2   Animation     7.914516   96.596774
3   Biography     7.862338  131.844156
4      Comedy     7.822436  107.602564
5       Crime     7.916935  122.298387
6       Drama     7.902518  126.539568
7      Family     7.850000  107.500000
8     Fantasy     7.700000  112.000000
9   Film-Noir     8.033333   97.333333
10    History     8.000000   66.000000
11     Horror     7.806897  102.517241
12    Mystery     7.975000  115.625000
13     Sci-Fi     7.920000  109.000000
14   Thriller     7.680000  114.200000
15    Western     8.255556  136.666667

##### = Western movies have the highest average running time, history has lowest